### ISSUE TITLE
Register allocation hints

### ISSUE THREAD

**Issue by mysterymath** [llvm-mos/llvm-mos#113]
Currently, the register allocator thinks that any legal location it can place a value is equally good. This isn't actually the case:

- ASL A is more efficient than ASL ZP
- Copying between A/X and A/Y is cheaper than copying between X/Y, except on the 65C02
- Coping between AXY/ZP is cheaper than copying ZP/ZP
- Etc.

To handle scenarios like this, LLVM provides a target hook to provide hints to the register allocator about where to place things. This logic can make decisions based on where things have already been placed, so most of the above should be achievable.

**Comment by beholdnec** [comment]
Which target hooks are you referring to? The only mechanism I found was the RegisterBankInfo::copyCost method, and we aren't really using RegisterBanks.

**Comment by mysterymath** [comment]
TargetRegisterInfo::getRegAllocationHints() and TargetRegisterInfo::updateRegAllocHint().

The former gets the initial list of hints for a virtual register, while the latter updates the hints if a copy is coalesced away.

**Comment by beholdnec** [comment]
I experimented with getRegAllocationHints. I don't think it will help in char-stats.ll's case. Here is my investigation.

The MOS::ASL instruction takes an AImag8 register. The register in question gets assigned to Imag8 (no A) very early in instruction selection, when a G_UADDE instruction is selected. See virtual register %60 below:

```
Selecting:
  %52:any(s8), %53:cc(s1) = G_UADDE %48:any, %60:any, %56:any
5: GIM_SwitchOpcode(MIs[0], [44, 189), Default=1273, JumpTable...) // Got=124
1274: GIM_Reject
1274: Rejected
Converting operand: %52:any
Converting operand: %53:cc
Converting operand: %85:_
Converting operand: %48:any(tied-def 0)
Converting operand: %60:any
Converting operand: %56:any
Into:
  %52:ac(s8), %53:cc(s1), %85:vc(s1) = ADCImag8 %48:ac(tied-def 0)(s8), %60:imag8(s8), %56:cc(s1)
```

Now %60 is assigned to an Imag8. Later, G_SHLE gets selected:

```
Selecting:
  %60:imag8(s8), %61:cc(s1) = G_SHLE %5:any, %56:cc
5: GIM_SwitchOpcode(MIs[0], [44, 189), Default=1273, JumpTable...) // Got=271
1274: GIM_Reject
1274: Rejected
Converting operand: %60:imag8
Converting operand: %61:cc
Converting operand: %5:any(tied-def 0)
Into:
  %60:imag8(s8), %61:cc(s1) = ASL %5:aimag8(tied-def 0)(s8)
```

Since %60 is already assigned to Imag8, this eventually forces the compiler to use the ASL_ZeroPage instruction.

Here is the final generated code, for reference:

```
.LBB0_2:                                ; %entry
                                        ; =>This Inner Loop Header: Depth=1
	jsr	next_char
	cmp	#0
	beq	.LBB0_5
; %bb.3:                                ; %while.body
                                        ;   in Loop: Header=BB0_2 Depth=1
	sta	mos8(__rc2)
	asl	mos8(__rc2)        ; shle here
	lda	#0
	sta	mos8(__rc3)
	rol	mos8(__rc3)
	clc
	lda	mos8(__rc0)
	ldx	mos8(__rc1)
	clc
	adc	mos8(__rc2)         ; add here
	sta	mos8(__rc2)
	txa
	adc	mos8(__rc3)
	sta	mos8(__rc3)
```

**Comment by mysterymath** [comment]
Yeah, this was a specific problem I ran into early on with ASL specifically. IIRC I had a sort of hack to deal with this, but I don't 100% remember what it was, and I don't think it generalized.

I don't think that
```%60:imag8(s8), %61:cc(s1) = ASL %5:aimag8(tied-def 0)(s8)```
is necessarily *wrong*, we actually do want the value %60 to end up in imaginary register.
It seems like we'd actually like the final sequence to be:
```
%60:aimag8(s8), %61:cc(s1) = ASL %5:aimag8(tied-def 0)(s8)
%61:imag8 = COPY %60
ADC ... %61:imag8
```

If we actually end up placing %60 in an imaginary register, then the register allocator should be able to coalesce away the copy. But we wouldn't want to preemptively coalesce this away, otherwise we lose the freedom to use the superior A register for the ASL.

If I recall correctly, way back when I was optimizing this originally, I tweaked the copy coalescer so that it would leave COPY operations like this alone. Not sure if one ever ends up being inserted in this case.

We also wouldn't want to insert such a copy if %5 was already an imag8. Then, there'd be no possible benefit.
